packages/apollo-tracing/src/index.ts

```diff
@@ -33,6 +33,97 @@ interface ResolverCall {
   endOffset?: HighResolutionTime;
 }
 
+export const plugin = (_futureOptions = {}) => (): ApolloServerPlugin => ({
+  requestDidStart() {
+    let startWallTime: Date | undefined;
+    let endWallTime: Date | undefined;
+    let startHrTime: HighResolutionTime | undefined;
+    let duration: HighResolutionTime | undefined;
+    const resolverCalls: ResolverCall[] = [];
+
+    startWallTime = new Date();
+    startHrTime = process.hrtime();
+
+    return {
+      executionDidStart() {
+        // It's a little odd that we record the end time after execution rather
+        // than at the end of the whole request, but because we need to include
+        // our formatted trace in the request itself, we have to record it
+        // before the request is over! It's also odd that we don't do traces for
+        // parse or validation errors, but runQuery doesn't currently support
+        // that, as format() is only invoked after execution.
+        return () => {
+          duration = process.hrtime(startHrTime);
+          endWallTime = new Date();
+        };
+      },
+      willResolveField(...args) {
+        const [, , , info] = args;
+
+        const resolverCall: ResolverCall = {
+          path: info.path,
+          fieldName: info.fieldName,
+          parentType: info.parentType,
+          returnType: info.returnType,
+          startOffset: process.hrtime(startHrTime),
+        };
+
+        resolverCalls.push(resolverCall);
+
+        return () => {
+          resolverCall.endOffset = process.hrtime(startHrTime);
+        };
+      },
+      willSendResponse({ response }) {
+        // In the event that we are called prior to the initialization of
+        // critical date metrics, we'll return undefined to signal that the
+        // extension did not format properly. Any undefined extension
+        // results are simply purged by the graphql-extensions module.
+        if (
+          typeof startWallTime === 'undefined' ||
+          typeof endWallTime === 'undefined' ||
+          typeof duration === 'undefined'
+        ) {
+          return;
+        }
+
+        const extensions =
+          response.extensions || (response.extensions = Object.create(null));
+
+        if (typeof extensions.tracing !== 'undefined') {
+          throw new Error("The tracing information already existed.");
```

Comment:
Possibly by another plugin? Curious what we're defending against here.

Reply:
Hopefully more clear with: https://github.com/apollographql/apollo-server/pull/3991/commits/7e4cc4b7d99aeaffe6a27b2cf97b98d7713c2a5c